airbyte-workers/src/main/java/io/airbyte/workers/storage/CloudDocumentStoreClient.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.workers.storage;
+
+import java.util.Optional;
+
+/**
+ * Interface for treating cloud storage like a simple document store.
+ */
+public interface CloudDocumentStoreClient {
```

Comment:
+1

Reply:
Could we avoid to catch exception and throw IO ones explicitly, In temporal we will add a catch block anyway in the activities that uses it.

Reply:
Yeah. I agree with this. I started to look into this and then decided it was a bad idea to block on that refactor because I think it is a hard refactor.

I think there are a lot of problems in how we are doing dependencies for the GCS. Namely the fact that we have made S3 and GCS a dependency of every module  (see slack convo [here](https://airbytehq.slack.com/archives/C0229LM09T4/p1638838002048800)). We should figure out how to consolidate all of the code around building these clients into a single module with tightly scope dependencies. That seems to interact in a weird way with our logging configuration which I don't fully grasp yet. I'm happy to look into do it in in my slack time, but it's not a small task and I don't think we should put it on the critical path for getting this scheduler project done.

Reply:
> Could we avoid to catch exception and throw IO ones explicitly, In temporal we will add a catch block anyway in the activities that uses it.

We definitely can. I'm trying to learn from our original `ConfigPersistence`. We did expose `IOException` there and then because it was such a pain to always be try / catching that exception in callers pretty much had `ConfigRepository` come into swallow the exceptions. If it meaningfully better to bubble up the exception here, we definitely can, but if it's just a style thing, then I think we save ourselves some headache by leaving it as is.

Reply:
> We definitely can. I'm trying to learn from our original `ConfigPersistence`. We did expose `IOException` there and then because it was such a pain to always be try / catching that exception in callers pretty much had `ConfigRepository` come into swallow the exceptions. If it meaningfully better to bubble up the exception here, we definitely can, but if it's just a style thing, then I think we save ourselves some headache by leaving it as is.

All the temporal activity will have their method wrapped in a try {} catch (Exception e). The goal is to be able to make a workflow to go into a quarantine state in case there is no point to re-run the sync if the exception won't be solve by a re-run (like a NPE).

Reply:
```suggestion
public interface DocumentStoreClient {
```

Reply:
agreed jared.

Reply:
discussed with @benmoriceau offline and we agreed to hold off on the IOException for now because the only implementation of the interface that throws one is the DockerCompose version. The others wouldn't, so adding it to the iface would be mostly misleading.